compiler/rustc_lint/src/types.rs

```diff
@@ -828,7 +828,7 @@ fn is_niche_optimization_candidate<'tcx>(
     match ty.kind() {
         ty::Adt(ty_def, _) => {
             let non_exhaustive = ty_def.is_variant_list_non_exhaustive();
-            let empty = (ty_def.is_struct() && ty_def.all_fields().next().is_none())
+            let empty = (ty_def.is_struct() && ty_def.non_enum_variant().fields.is_empty())
                 || (ty_def.is_enum() && ty_def.variants().is_empty());
```

Comment:
To document why this is OK... `non_enum_variant` asserts that `self` is either a struct or a union; we've just tested that this is a struct.  It then returns its unique variant, and we test whether its fields are empty.  That's equivalent to constructing an iterator that walks over all fields of all variants, which is what `all_fields` does, and testing there aren't any.